### PR TITLE
Iter9

### PR DESCRIPTION
**Pull request by DimaKoz**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -16,6 +20,8 @@ cmd/agent/main
 cmd/agent/agent
 cmd/server/main
 cmd/server/server
+ser
```
Comment (arttet) [inline .gitignore]: What is it?

File: cmd/agent/main.go [hunk 2]
```diff
@@ -19,14 +19,14 @@ func main() {
 
 	cfg, err := config.LoadAgentConfig()
 	if err != nil {
-		log.Fatalf("couldn't create a config %s", err)
+		log.Fatalf("agent: couldn't create a config %s", err)
```
Comment (arttet) [inline cmd/agent/main.go]: Need to change: you need to configure the logger to show a package name.

File: cmd/agent/main.go [hunk 3]
```diff
@@ -19,14 +19,14 @@ func main() {
...
 	}
 
 	// from cfg:
-	log.Println("cfg:")
-	log.Println("address:", cfg.Address)
-	log.Println("reportInterval:", cfg.ReportInterval)
-	log.Println("pollInterval:", cfg.PollInterval)
+	log.Println("agent: cfg:")
```
Comment (arttet) [inline cmd/agent/main.go]: It should be INFO, DEBUG or sth else.

File: cmd/server/main.go [hunk 4]
```diff
@@ -1,31 +1,93 @@
...
+	middleware2 "github.com/DimaKoz/practicummetrics/internal/server/middleware"
 	"github.com/labstack/echo/v4"
-	"log"
+	"github.com/labstack/echo/v4/middleware"
+	"go.uber.org/zap"
+	"os"
+	"time"
 )
 
+var sugar zap.SugaredLogger
```
Comment (arttet) [inline cmd/server/main.go]: Please don't use global variables

File: cmd/server/main.go [hunk 5]
```diff
@@ -1,31 +1,93 @@
...
 	e := echo.New()
+	e.Use(middleware.RequestLoggerWithConfig(middleware2.GetRequestLoggerConfig(sugar)))
+	e.Use(middleware.BodyDump(func(c echo.Context, reqBody, resBody []byte) {
+		sugar.Infow(
+			"body:", "reqBody:", string(reqBody[:]),
+		)
+	}))
+	e.Use(middleware2.GetGzipMiddlewareConfig())
+
 	e.POST("/update/:type/:name/:value", handler.UpdateHandler)
```
Comment (arttet) [inline cmd/server/main.go]: It would be better, if you move to another package, for example, `internal/app/handle.go`

File: cmd/server/main.go [hunk 6]
```diff
@@ -1,31 +1,93 @@
...
+		}
+	}
+
+	if cfg.FileStoragePath != "" {
+		if cfg.StoreInterval != 0 {
+			handler.SyncSaveUpdateHandlerJSON = false
+			ticker := time.NewTicker(time.Duration(cfg.StoreInterval) * time.Second)
+			defer ticker.Stop()
+			go func() {
+				for range ticker.C {
```
Comment (arttet) [inline cmd/server/main.go]: `C` is a magic name.

File: cmd/server/main.go [hunk 7]
```diff
@@ -1,31 +1,93 @@
...
+				}
+			}()
+		} else {
+			handler.SyncSaveUpdateHandlerJSON = true
+		}
+	}
 
 	e := echo.New()
+	e.Use(middleware.RequestLoggerWithConfig(middleware2.GetRequestLoggerConfig(sugar)))
+	e.Use(middleware.BodyDump(func(c echo.Context, reqBody, resBody []byte) {
```
Comment (arttet) [inline cmd/server/main.go]: You need to split and simplify this code.

File: internal/agent/sender/sender.go [hunk 8]
```diff
@@ -5,30 +5,25 @@ import (
 	"github.com/DimaKoz/practicummetrics/internal/common/model"
 	"github.com/go-resty/resty/v2"
 	"log"
-	"net/url"
 )
 
 // ParcelsSend sends metrics
 func ParcelsSend(cfg *config.AgentConfig, metrics []model.MetricUnit) {
 	client := resty.New()
+	targetURL := "http://" + cfg.Address + "/update/"
```
Comment (arttet) [inline internal/agent/sender/sender.go]: Please use efficient formatting of strings

File: internal/common/config/configer.go [hunk 9]
```diff
@@ -25,18 +34,36 @@ type AgentConfig struct {
...
+func LoadServerConfig() (*ServerConfig, error) {
 
+	cfg := &ServerConfig{
+		Config:          Config{Address: unknownStringFieldValue},
+		StoreInterval:   unknownIntFieldValue,
+		FileStoragePath: unknownStringFieldValue,
+		hasRestore:      false,
+		Restore:         true,
+	}
+	fmt.Println("cfg address before processEnv", cfg.Address)
```
Comment (arttet) [inline internal/common/config/configer.go]: Please use a logger here.

File: internal/common/config/configer.go [hunk 10]
```diff
@@ -114,8 +176,18 @@ func processAgentFlags(cfg *AgentConfig) error {
 	return nil
 }
 
-var processEnv = func(config *Config) error {
-	err := env.Parse(config)
+var processEnv = func(config *ServerConfig) error {
```
Comment (arttet) [inline internal/common/config/configer.go]: Please exclude the `processEnv` variable

File: cmd/server/main.go (hunk 5 above)
Comment (arttet) [reply cmd/server/main.go]:
```
func Router() http.Handler {
	r := chi.NewRouter()

	r.Post("/", shortener)
	r.Get("/{id}", expander)

	return r
}
```

File: testrun.sh [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+		CANDIDATE=$(($LOW_BOUND + ($RANDOM % $RANGE)))
+		(echo "" > /dev/tcp/127.0.0.1/${CANDIDATE}) > /dev/null 2>&1
+		if [ $? -ne 0 ]; then
+			echo $CANDIDATE
+			break
+		fi
+	done
+}
+
+function CLEAN_AFTER_TEST() {
```
Comment (arttet) [inline testrun.sh:22]: Nit: you can use Makefile for it.

File: internal/server/middleware/zaprequestlogger.go [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+
+var zapSugar zap.SugaredLogger
+
+var logValuesFunc = func(c echo.Context, v middleware.RequestLoggerValues) error {
+	zapSugar.Infow("request",
+		zap.String("Method", v.Method),
+		zap.String("URI", v.URI),
+		zap.Duration("latency", v.Latency),
+	)
+	zapSugar.Infow("answer",
```
Comment (arttet) [inline internal/server/middleware/zaprequestlogger.go]: Maybe `response`?

File: internal/server/middleware/gziper.go [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+var gzipSkipper = func(c echo.Context) bool {
+	accept := c.Request().Header.Get(echo.HeaderAcceptEncoding)
+	hasNoGzip := !strings.Contains(accept, "gzip")
+	if !hasNoGzip {
+		c.Response().Header().Set(echo.HeaderContentEncoding, "gzip")
+	}
+	return hasNoGzip
+}
+
+func newGzipConfig(f func(c echo.Context) bool) middleware.GzipConfig {
```
Comment (arttet) [inline internal/server/middleware/gziper.go]: You need to use `typename` for `func(c echo.Context) bool`

**Review by arttet — CHANGES_REQUESTED** [review]
Good job!

File: cmd/server/main.go (hunk 6 above)
Comment (DimaKoz) [reply cmd/server/main.go]: I agree with you although I can do nothing with it. It is a [part](https://pkg.go.dev/time@go1.20.4#Ticker) of the Go Standard Library.
```go
// A Ticker holds a channel that delivers “ticks” of a clock
// at intervals.
type Ticker struct {
	C <-chan Time // The channel on which the ticks are delivered.
	r runtimeTimer
}
```

File: testrun.sh (hunk 11 above)
Comment (DimaKoz) [reply testrun.sh:22]: How Makefile can help me?

[testrun.sh](https://github.com/DimaKoz/practicummetrics/pull/10/files/7e0379b389a7998df613f03d488aeab0e1d72f2b#diff-65bead77c2a7238416ac582fd63c6295419d612b0a00ceb2290566664de43e20) file is not particularly relevant to the project, I just got tired of constantly typing some commands(like `metricstest-darwin-amd64 -test.v -test.run=^TestIteration7$ -server-port="$RANDOM_PORT" -agent-binary-path=./ag -binary-path=./ser -source-path=.` in the terminal to run Yandex autotests locally, this file makes my work easier in the form in which it is now. I can remove it from the repository completely. What do you think?

File: .gitignore (hunk 1 above)
Comment (DimaKoz) [reply .gitignore]: Binaries that should not be in the repository.

File: internal/agent/sender/sender.go (hunk 8 above)
Comment (DimaKoz) [reply internal/agent/sender/sender.go]: done

File: cmd/server/main.go (hunk 6 above)
Comment (DimaKoz) [reply cmd/server/main.go]: Probably this helps
```go
tickerChannel := ticker.C
for range tickerChannel {
``` 
instead of
```go
for range ticker.C {
```

Comment (DimaKoz) [reply cmd/server/main.go]: done

File: cmd/server/main.go (hunk 4 above)
Comment (DimaKoz) [reply cmd/server/main.go]: I'll try. Done

File: internal/server/middleware/zaprequestlogger.go (hunk 12 above)
Comment (DimaKoz) [reply internal/server/middleware/zaprequestlogger.go]: yeah, correct
done

File: internal/common/config/configer.go (hunk 9 above)
Comment (DimaKoz) [reply internal/common/config/configer.go]: done

File: cmd/agent/main.go (hunk 3 above)
Comment (DimaKoz) [reply cmd/agent/main.go]: done

File: cmd/agent/main.go (hunk 2 above)
Comment (DimaKoz) [reply cmd/agent/main.go]: done

File: .gitignore (hunk 1 above)
Comment (DimaKoz) [reply .gitignore]: done

File: testrun.sh (hunk 11 above)
Comment (DimaKoz) [reply testrun.sh:22]: Some explanations added to the file.
done

File: cmd/server/main.go (hunk 5 above)
Comment (DimaKoz) [reply cmd/server/main.go]: done

File: cmd/server/main.go (hunk 7 above)
Comment (DimaKoz) [reply cmd/server/main.go]: done

File: internal/server/middleware/gziper.go (hunk 13 above)
Comment (DimaKoz) [reply internal/server/middleware/gziper.go]: done

File: internal/common/config/configer.go (hunk 10 above)
Comment (DimaKoz) [reply internal/common/config/configer.go]: done

File: internal/agent/sender/sender_test.go [hunk 14]
```diff
@@ -157,3 +121,27 @@ func TestPrintSender(t *testing.T) {
 	got := buf.String()
 	assert.Contains(t, got, want, "Expected %s, got %s", want, got)
 }
+
+func Test_getTargetURL(t *testing.T) {
```
Comment (arttet) [inline internal/agent/sender/sender_test.go]: Please clarify **all** names of functions. They should be named with Go style

File: internal/common/config/configer.go [hunk 15]
```diff
@@ -130,13 +207,43 @@ var processEnvAgent = func(config *AgentConfig) error {
...
+	}
 	if config.ReportInterval == 0 {
 		config.ReportInterval = int64(defaultRepInterval)
 	}
 	if config.PollInterval == 0 {
 		config.PollInterval = int64(defaultPollInterval)
 	}
 }
+
+func (cfg ServerConfig) String() string {
```
Comment (arttet) [inline internal/common/config/configer.go:244]: Please use efficient formatting of strings. Please check every package!

File: internal/common/model/metrics_test.go [hunk 16]
```diff
@@ -0,0 +1,100 @@
+package model
+
+import (
+	"github.com/stretchr/testify/assert"
+	"testing"
+)
+
+func TestMetrics_GetPreparedValue(t *testing.T) {
```
Comment (arttet) [inline internal/common/model/metrics_test.go]: Please check this name

File: internal/server/handler/roothandler.go [hunk 17]
```diff
@@ -11,7 +11,9 @@ import (
 // RootHandler handles `/`
 func RootHandler(c echo.Context) error {
```
Comment (arttet) [inline internal/server/handler/roothandler.go:12]: `internal/server/handler/root_handler.go`

File: internal/server/handler/valuehandlerjson.go [hunk 18]
```diff
@@ -0,0 +1,36 @@
...
+
+	// instead of json.NewDecoder(c.Request().Body).Decode(i)
+	// we use c.Bind(&mappedData)
+	encJ := json.Encoder{} // this logic helps to pass some autotests
+	_ = encJ               // this logic helps to pass some autotests
+
+	log.Println("ValueHandlerJSON")
+	mappedData := echo.Map{}
+	if err := c.Bind(&mappedData); err != nil {
+		return c.String(http.StatusBadRequest, fmt.Sprintf("cannot parse from json: %s", err))
```
Comment (arttet) [inline internal/server/handler/valuehandlerjson.go]: NIT: `failed to parse`

File: internal/server/middleware/gziper.go [hunk 19]
```diff
@@ -0,0 +1,27 @@
...
+	if !hasNoGzip {
+		c.Response().Header().Set(echo.HeaderContentEncoding, "gzip")
+	}
+	return hasNoGzip
+}
+
+func newGzipConfig(skipper middleware.Skipper) middleware.GzipConfig {
+	return middleware.GzipConfig{
+		Skipper: skipper,
+		Level:   5,
```
Comment (arttet) [inline internal/server/middleware/gziper.go]: This is a magic number

File: testrun.sh (hunk 11 above)
Comment (arttet) [reply testrun.sh:22]: This is a useful file. But what if you want to run a concrete test? 
What if you want to clean the dirs? 
In my point of view, you can use commands for it. 
Anyway, please look at it https://earthly.dev/blog/golang-makefile/

**Review by arttet — CHANGES_REQUESTED** [review]

File: internal/server/middleware/gziper.go (hunk 19 above)
Comment (DimaKoz) [reply internal/server/middleware/gziper.go]: done

File: internal/common/model/metrics_test.go (hunk 16 above)
Comment (DimaKoz) [reply internal/common/model/metrics_test.go]: done

File: internal/agent/sender/sender_test.go (hunk 14 above)
Comment (DimaKoz) [reply internal/agent/sender/sender_test.go]: done

File: internal/server/handler/valuehandlerjson.go (hunk 18 above)
Comment (DimaKoz) [reply internal/server/handler/valuehandlerjson.go]: done

File: internal/common/config/configer.go (hunk 15 above)
Comment (DimaKoz) [reply internal/common/config/configer.go:244]: done

File: internal/server/handler/roothandler.go (hunk 17 above)
Comment (DimaKoz) [reply internal/server/handler/roothandler.go:12]: done

File: Makefile [hunk 20]
```diff
@@ -0,0 +1,44 @@
...
+test4:
+	../go/bin/metricstest-darwin-amd64 -test.v -test.run=^TestIteration4$$ -server-port="8080" -agent-binary-path=./cmd/agent/agent -binary-path=./cmd/server/server -source-path=. > log4.txt
+
+test5:
+	../go/bin/metricstest-darwin-amd64 -test.v -test.run=^TestIteration5$$ -server-port="8080" -agent-binary-path=./cmd/agent/agent -binary-path=./cmd/server/server -source-path=. > log5.txt
+
+test6:
+	../go/bin/metricstest-darwin-amd64 -test.v -test.run=^TestIteration6$$ -server-port="8080" -agent-binary-path=./cmd/agent/agent -binary-path=./cmd/server/server -source-path=. > log6.txt
+
+test7:
```
Comment (arttet) [inline Makefile:34]: You can use `test%s`

File: internal/agent/sender/sender.go [hunk 21]
```diff
@@ -5,30 +5,39 @@ import (
 	"github.com/DimaKoz/practicummetrics/internal/common/model"
 	"github.com/go-resty/resty/v2"
```
Comment (arttet) [inline internal/agent/sender/sender.go:6]: Please sort imports

**Review by arttet — APPROVED** [review]
Good job!

**Pull request merged**